src/Microsoft.Extensions.Logging.Abstractions/ISupportsExternalScope.cs

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Logging
+{
+    /// <summary>
+    /// An interface for configuring logging providers.
+    /// </summary>
+    public interface ISupportsExternalScope
+    {
+        void SetScopeProvider(IExternalScopeProvider scopeProvider);
+    }
+
+    public interface IExternalScopeProvider
+    {
+        void CollectScope<T>(Action<object, T> callback, T state);
+        IDisposable Push(object state);
```

Comment:
Since BeginScope takes `TState state`, shouldn't this too?

Reply:
`Push<TState>(TState state)`?

This is one place where the external scope provider isn't able to have the advantages that each provider has. We preserve generics all the way through in order to allow `structs` to remain unboxed, but the external provider will be forced to box them in order to store them for later.